params/protocol_params.go

```diff
@@ -162,9 +172,61 @@ func (t CodeFormat) String() string {
 	return "UndefinedCodeFormat"
 }
 
-func (t CodeFormat) Validate() bool {
-	if t < CodeFormatLast {
-		return true
+// VmVersion contains the information of the contract deployment time (ex. 0x0(constantinople), 0x1(istanbul,...))
+type VmVersion uint8
+
+// Supporting VmVersion
+const (
+	VmVersion0 VmVersion = iota // Deployed at Constantinople
+	VmVersion1                  // Deployed at Istanbul, ...(later HFs would be added)
+)
+
+func (t VmVersion) String() string {
+	return "VmVersion" + string(t)
+}
+
+// CodeInfo consists of 8 bits, and has information of the contract code.
+// Originally, codeInfo only contains codeFormat information(interpreter version), but now it is divided into two parts.
+// First four bit contains the deployment time (ex. 0x00(constantinople), 0x10(istanbul,...)), so it is called vmVersion.
+// Last four bit contains the interpreter version (ex. 0x00(EVM), 0x01(EWASM)), so it is called codeFormat.
+type CodeInfo uint8
+
+const (
+	// codeFormatBitMask filters only the codeFormat. It means the interpreter version used by the contract.
+	// Mask result 1. [x x x x 0 0 0 1]. The contract uses EVM interpreter.
+	codeFormatBitMask = 0b00001111
+
+	// vmVersionBitMask filters only the vmVersion. It means deployment time of the contract.
+	// Mask result 1. [0 0 0 0 x x x x]. The contract is deployed at constantinople
+	// Mask result 2. [0 0 0 1 x x x x]. The contract is deployed after istanbulHF
+	vmVersionBitMask = 0b11110000
+)
+
+func NewCodeInfo(codeFormat CodeFormat, vmVersion VmVersion) CodeInfo {
+	return CodeInfo(codeFormat) | CodeInfo(vmVersion)<<4
```

Comment:
It's not necessary but I think masking for codeFormat is more safe wasy
```suggestion
	return CodeInfo(codeFormat & codeFormatBitMask) | CodeInfo(vmVersion)<<4
```

Reply:
Agreed that it is not necessary for now, but it is more safe.. thanks for the suggestion!